staging/src/k8s.io/component-base/config/testing/defaulting.go

```diff
@@ -38,9 +37,9 @@ func GetDefaultingTestCases(t *testing.T, scheme *runtime.Scheme, codecs seriali
 		if gvk.Version == runtime.APIVersionInternal {
 			continue
 		}
-		beforeDir := fmt.Sprintf("testdata/%s/before", gvk.Kind)
```

Comment:
I *really* don't want to see us double-down on OS-specific API defaulting... that makes the content of a serialized API config file require knowledge of the target OS environment to round-trip properly.

If we have OS-specific code paths, that should not be using the API-defaulting mechanic, it should be a runtime behavior during kubelet setup based on the absence of a field.

Reply:
I took another look at this. I think it's already behaving as you say. The defaults are being set during runtime, based on the absence of a field in the kubelet configuration file.

So, for the defaulting test cases, the tests basically receives the ``before`` KubeletConfiguration file, which looks something like this:

```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
```

As you can see, there's nothing OS-specific here. The Kubelet default config values are then applied over this, and then the tests assert that the expected configuration matches the actual configuration which is then going to be used by Kubelet, based on what defaults are being set in ``pkg/kubelet/apis/config/v1beta1/defaults.go``. In there, we have this: https://github.com/kubernetes/kubernetes/blob/6dbec8e25592d47fc8a8269c86d4b5fa838d320b/pkg/kubelet/apis/config/v1beta1/defaults.go#L211-L213 . As you can see, if the ``EvictionHard`` is not set, the ``DefaultEvictionHard`` is set, which has 2 different values, based on the OS: [1] [2]

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/v1beta1/defaults_linux.go
[2] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/v1beta1/defaults_others.go

This is the source of the failure on Windows / darwin, the test expects the Linux defaults to be applied in the Windows / darwin case as well. This PR basically adds the expected default configuration for the Windows case (although, we could add ``_other.yaml`` instead of just ``_windows.yaml``, wdyt?).

One interesting thing to note outside the scope of this PR, is that a user could have the ``"nodefs.inodesFree": "5%",`` in the ``EvictionHard`` section of the kubelet file on a Windows node, which is not sanitized by the runtime. I'm not sure what this can lead to, but it might be something worth investigating.

Reply:
> I took another look at this. I think it's already behaving as you say. The defaults are being set during runtime, based on the absence of a field in the kubelet configuration file.

I mean set at runtime, distinct from decoding the config file.

Today, this is the code path:
* decode v1beta1 config file → apply API defaults (OS-specific) → convert to internal kubelet config struct → validate → start kubelet

OS-specific defaults mean we can't round-trip a config file without baking in defaults for the OS the round-tripping was done on. This should be the path:
* decode v1beta1 config file → apply API defaults (non-OS-specific) → convert to internal kubelet config struct → validate → start kubelet → handle OS-specific default behavior for unspecified values

Reply:
@liggitt, I've updated the PR. The ``EvictionHard`` options (which contains the OS-specific fields) are now set after the kubelet config struct is generated.